File: leetcode/103-binary-tree-zigzag-level-order-traversal.js
// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/ (top-interview-questions)


// Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}


/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
  const levels = [];
  const queue = root !== null ? [root] : [];
  
  while (queue.length > 0) {
    const levelLength = queue.length;
    levels.push([]);
    for (let i = 0; i < levelLength; i++) {
      const node = queue.pop();
      levels[levels.length - 1].push(node.val);
      if (node.left) queue.unshift(node.left);
      if (node.right) queue.unshift(node.right);
    }

    if (levels.length % 2 === 0) {
      // if the level number is pair we should reverse it value order
      levels[levels.length - 1].reverse();
    }
  }

  return levels;
};

File: leetcode/103-binary-tree-zigzag-level-order-traversal.py
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/ (top-interview-questions)


from collections import deque
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        values = []

        queue = deque([root] if root else [])
        
        while queue:
            level_values = []
            level_length = len(queue)
            for i in range(level_length):
                node = queue.popleft()
                level_values.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        
            values.append(level_values if len(values) % 2 == 0 else reversed(level_values))
        
        return values
